secant/Features/BackupFlow/RecoveryPhraseValidation.swift

```diff
@@ -0,0 +1,229 @@
+            let missingIndex = missingIndices[i]
+            // figure out what this means in terms of the whole recovery phrase
+            let concreteIndex = i * groupLength + missingIndex
+            precondition(concreteIndex < words.count)
+            // replace the word on the copy of the original phrase with the completion the user did
+            words[concreteIndex] = wordCompletion.word
+        }
+
+        return words
+    }
+
+    /// drives the state machine represented on this Enum by the action of applying a chip into a group of words containing an empty slot that has to be completed
+    static func given(_ state: RecoveryPhraseValidationState, apply chip: PhraseChip.Kind, into group: Int) -> Self {
+        guard case let PhraseChip.Kind.unassigned(word) = chip else { return state }
+
+        switch state.step {
+        case .initial:
+            guard let missingChipIndex = state.missingWordChips.firstIndex(of: chip) else { return state }
+
+            var newMissingWords = state.missingWordChips
+            newMissingWords[missingChipIndex] = .empty
+
+            return RecoveryPhraseValidationState(
+                phrase: state.phrase,
+                missingIndices: state.missingIndices,
+                missingWordChips: newMissingWords,
+                completion: [RecoveryPhraseStepCompletion(groupIndex: group, word: word)]
+            )
+
+        case .incomplete:
+            guard let missingChipIndex = state.missingWordChips.firstIndex(of: chip) else { return state }
+
+            if state.completion.count < (RecoveryPhraseValidationState.phraseChunks - 1) {
+                var newMissingWords = state.missingWordChips
+                newMissingWords[missingChipIndex] = .empty
+
+                var newCompletionState = Array(state.completion)
+                newCompletionState.append(RecoveryPhraseStepCompletion(groupIndex: group, word: word))
+
+                return RecoveryPhraseValidationState(
+                    phrase: state.phrase,
+                    missingIndices: state.missingIndices,
+                    missingWordChips: newMissingWords,
+                    completion: newCompletionState
+                )
+            } else {
+                var newCompletion = state.completion
+                newCompletion.append(RecoveryPhraseStepCompletion(groupIndex: group, word: word))
+
+                return RecoveryPhraseValidationState(
+                    phrase: state.phrase,
+                    missingIndices: state.missingIndices,
+                    missingWordChips: Array(repeating: .empty, count: RecoveryPhraseValidationState.phraseChunks),
+                    completion: newCompletion
+                )
+            }
+        default:
+            return state
+        }
+    }
+
+    static func pickWordsFromMissingIndices(indices: [Int], phrase: RecoveryPhrase) -> [PhraseChip.Kind] {
```

Comment:
I don't think any of these functions need to be static, everything appears to be able to be inlined (doesn't need a separate function) or they can all be a function that's a part of some Value, making it easier to read and reason about, and less properties to pass around.

Reply:
This one is used on constructors. So it must be something external to the instance being created

Reply:
Can it just be used inline? Its only used once.  If you want a separate function, should this live on the RecoverPhrase so you don't have to pass it in and it can be a not-static function?

Reply:
not quite so because the helper function is used to create the test scenarios and other helper functions that need this behavior to be outside the struct.